File: src/problems_all/202_happy_number/202_happy_number_v2.go
/*
https://leetcode.com/problems/happy-number/description/

Runtime: 0 ms
Memory: 2 MB
https://leetcode.com/problems/happy-number/submissions/881273371/
*/

package main

import "fmt"

func main() {
	n := 7
	result := isHappy(n)
	fmt.Println(result)
}

func sumOfDigits(n int) int {
	total := 0
	for n > 0 {
		digit := n % 10
		total += (digit * digit)
		n /= 10
	}
	return total
}

func isHappy(n int) bool {
	if n == 1 {
		return true
	}

	happy := true
	seq := map[int]bool{}
	for {
		total := sumOfDigits(n)

		if total == 1 {
			break
		}

		if _, ok := seq[total]; ok {
			happy = false
			break
		}

		seq[total] = true
		n = total
	}

	return happy
}

File: src/problems_all/202_happy_number/202_happy_number_v2.py
"""
https://leetcode.com/problems/happy-number/description/

Runtime: 36 ms
Memory: 14 MB
https://leetcode.com/problems/happy-number/submissions/318642685/
"""


class Solution:
    def isHappy(self, n: int) -> bool:
        if n == 1:
            return True
            
        seq = set()
        while True:
            nums = [(int(x) ** 2) for x in str(n)]
            total = sum(nums)
            
            if total == 1:
                return True

            if total in seq:
                return False
            
            seq.add(total)
            n = total

        return False


def main():
    n = 2
    result = Solution().isHappy(n)
    print(result)


if __name__ == '__main__':
    main()
